source/bluetooth/MicroBitPartialFlashingService.cpp

```diff
@@ -0,0 +1,322 @@
+    partialFlashCharacteristic.requireSecurity(SecurityManager::MICROBIT_BLE_SECURITY_LEVEL);
+
+    // Create Partial Flashing Service
+    GattCharacteristic *characteristics[] = {&partialFlashCharacteristic};
+    GattService         service(MicroBitPartialFlashingServiceUUID, characteristics, sizeof(characteristics) / sizeof(GattCharacteristic*) );
+    ble.addService(service);
+
+    // Get characteristic handle for future use
+    partialFlashCharacteristicHandle = partialFlashCharacteristic.getValueHandle();
+    ble.gattServer().onDataWritten(this, &MicroBitPartialFlashingService::onDataWritten);
+
+    // Set up listener for SD writing
+    messageBus.listen(MICROBIT_ID_PFLASH_NOTIFICATION, MICROBIT_EVT_ANY, this, &MicroBitPartialFlashingService::partialFlashingEvent);
+
+    // Set up Memory map
+    // This will create the Memory Map and store it in flash
+    MicroBitMemoryMap memoryMap;
+
+}
+
+
+/**
+  * Callback. Invoked when any of our attributes are written via BLE.
+  */
+void MicroBitPartialFlashingService::onDataWritten(const GattWriteCallbackParams *params)
+{
+    // Get data from BLE callback params
+    uint8_t *data = (uint8_t *)params->data;
+
+    if(params->handle == partialFlashCharacteristicHandle && params->len > 0)
+    {
+      // Switch CONTROL byte
+      switch(data[0]){
+        case REGION_INFO:
+        {
+          // Create instance of Memory Map to return info
+          MicroBitMemoryMap memoryMap;
+
+          uint8_t buffer[18];
+          // Response:
+          // Region and Region #
+          buffer[0] =  0x00;
+          buffer[1] =  data[1];
+
+          // Start Address
+          buffer[2] = (memoryMap.memoryMapStore.memoryMap[data[1]].startAddress & 0xFF000000) >> 24;
+          buffer[3] = (memoryMap.memoryMapStore.memoryMap[data[1]].startAddress & 0x00FF0000) >> 16;
+          buffer[4] = (memoryMap.memoryMapStore.memoryMap[data[1]].startAddress & 0x0000FF00) >>  8;
+          buffer[5] = (memoryMap.memoryMapStore.memoryMap[data[1]].startAddress & 0x000000FF);
+
+          // End Address
+          buffer[6] = (memoryMap.memoryMapStore.memoryMap[data[1]].endAddress & 0xFF000000) >> 24;
+          buffer[7] = (memoryMap.memoryMapStore.memoryMap[data[1]].endAddress & 0x00FF0000) >> 16;
+          buffer[8] = (memoryMap.memoryMapStore.memoryMap[data[1]].endAddress & 0x0000FF00) >>  8;
+          buffer[9] = (memoryMap.memoryMapStore.memoryMap[data[1]].endAddress & 0x000000FF);
+
+          // Region Hash
+          for(int i = 0; i < 8; ++i)
+              buffer[10+i] = memoryMap.memoryMapStore.memoryMap[data[1]].hash[i];
+
+          // Send BLE Notification
+          ble.gattServer().notify(partialFlashCharacteristicHandle, (const uint8_t *)buffer, 18);
+
+          // Set offset for writing
+          baseAddress = (memoryMap.memoryMapStore.memoryMap[data[1]].startAddress & 0xFFFF0000) >> 16; // Offsets are 16 bit
+
+          // Reset packet count
+          packetCount = 0;
```

Comment:
@microbit-sam Looks like we also initialise blockPacketCount, blockNum and offset to zero statically when an instance of this class is created... do we need to reset any of this state here too?